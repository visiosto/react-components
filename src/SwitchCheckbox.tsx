// Copyright (c) 2021 Visiosto oy
// Licensed under the MIT License

import React from 'react';
import styled from 'styled-components';

import switchInputStyle from './style/switchInputStyle';
import switchLabelStyle from './style/switchLabelStyle';
import switchSpanStyle from './style/switchSpanStyle';

import { formClassErrorMessage } from './constants';

const SwitchInputSpan = styled.span`
  ${switchLabelStyle}

  input {
    ${switchInputStyle}
  }
`;

const SwitchSpan = styled.span`
  ${switchSpanStyle}
`;

export type SwitchCheckboxProps = {
  checked?: boolean;
  errorMessage?: string;
  handleClick: React.MouseEventHandler | undefined;
  id: string;
  label: string;
  name: string;
};

function SwitchCheckbox({
  checked = false,
  errorMessage = '',
  handleClick,
  id,
  label,
  name,
}: SwitchCheckboxProps) {
  return (
    <>
      <label htmlFor={id}>{label}</label>
      {errorMessage !== '' ? (
        <label className={formClassErrorMessage} htmlFor={id}>
          {errorMessage}
        </label>
      ) : null}
      <SwitchInputSpan onClick={handleClick}>
        <input key={Math.random()} defaultChecked={checked} id={id} name={name} type="checkbox" />
        <SwitchSpan />
      </SwitchInputSpan>
    </>
  );
}

export default SwitchCheckbox;
