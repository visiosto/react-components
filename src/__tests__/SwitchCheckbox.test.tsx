// Copyright (c) 2021 Visiosto oy
// Licensed under the MIT License

import React from 'react';
import { screen } from '@testing-library/react';

import SwitchCheckbox from '../SwitchCheckbox';

import renderWithProviders from '../util/test/renderWithProviders';

describe('Switch checkbox component', () => {
  it('renders correctly', () => {
    const { container } = renderWithProviders(
      <SwitchCheckbox
        handleClick={jest.fn()}
        id="switch-checkbox"
        label="Switch checkbox"
        name="switch-checkbox"
        checked
      />,
    );

    expect(container).toMatchSnapshot();

    expect(screen.getByLabelText('Switch checkbox')).toBeChecked();
  });

  it('renders correctly when not checked', () => {
    renderWithProviders(
      <SwitchCheckbox
        checked={false}
        handleClick={jest.fn()}
        id="switch-checkbox"
        label="Switch checkbox"
        name="switch-checkbox"
      />,
    );

    expect(screen.getByLabelText('Switch checkbox')).not.toBeChecked();
  });
});
