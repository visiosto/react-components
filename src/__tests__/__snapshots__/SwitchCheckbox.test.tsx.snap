// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch checkbox component renders correctly 1`] = `
<div>
  <label
    for="switch-checkbox"
  >
    Switch checkbox
  </label>
  <span
    class="SwitchCheckbox__SwitchInputSpan-sc-1vntclg-0 cpHXSd"
  >
    <input
      checked=""
      id="switch-checkbox"
      name="switch-checkbox"
      type="checkbox"
    />
    <span
      class="SwitchCheckbox__SwitchSpan-sc-1vntclg-1 gPJtgD"
    />
  </span>
</div>
`;
